arch/arm/src/samv7/sam_lin_sock.c

```diff
@@ -0,0 +1,1503 @@
+
+  priv->dev.d_buf = (uint8_t *)priv->txdesc;
+
+  if (!priv->master)
+    {
+      /* Enabled interrupts for LIN slave so we can wait for LIN ID */
+
+      sam_lin_putreg(priv, SAM_UART_IER_OFFSET,
+                     UART_INT_LINID | UART_INT_LINERR);
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Function: sam_lin_ifdown
+ *
+ * Description:
+ *   NuttX Callback: Stop the interface.
+ *
+ * Input Parameters:
+ *   dev  - Reference to the NuttX driver state structure
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+static int sam_lin_ifdown(struct net_driver_s *dev)
+{
+  struct sam_lin_s *priv = (struct sam_lin_s *)dev->d_private;
+
+  /* Disable LIN interrupts */
+
+  sam_lin_shutdown(priv);
+
+  /* Reset LIN */
+
+  sam_lin_reset(priv);
+
+  if (!priv->master)
+    {
+      int i;
+
+      /* Discard all cached data */
+
+      for (i = 0; i < LIN_ID_MASK + 1; i++)
+        {
+          priv->tx_cache[i].can_id = 0;
+        }
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: sam_lin_txready
+ *
+ * Description:
+ *   Return true if the LIN hardware can accept another TX message.
+ *
+ ****************************************************************************/
+
+static bool sam_lin_txready(struct sam_lin_s *priv)
+{
+  if (priv->master)
+    {
+      uint32_t regval;
+
+      /* Return true no interrupts are enabled */
+
+      regval = sam_lin_getreg(priv, SAM_UART_IMR_OFFSET);
+      ninfo("LIN%" PRIu8 " IMR: %08" PRIx32 "\n", priv->port, regval);
+
+      return regval == 0;
+    }
+  else
+    {
+      /* Slave is always ready for transmission since all
+       * TX data are cached
+       */
+
+      return true;
+    }
+}
+
+/****************************************************************************
+ * Name: sam_lin_transmit
+ *
+ * Description:
+ *   Start hardware transmission.  Called either from the txdone interrupt
+ *   handling or from watchdog based polling.
+ *
+ * Input Parameters:
+ *   priv - reference to the private LIN driver state structure
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure
+ *
+ * Assumptions:
+ *   May or may not be called from an interrupt handler.  In either case,
+ *   global interrupts are disabled, either explicitly or indirectly through
+ *   interrupt handling logic.
+ *
+ ****************************************************************************/
+
+static int sam_lin_transmit(struct sam_lin_s *priv)
+{
+  struct can_frame *frame = (struct can_frame *)priv->dev.d_buf;
+  uint32_t          regval;
+  uint8_t           lin_id;
+
+  ninfo("LIN%" PRIu8 " ID: %" PRIu32 " DLC: %" PRIu8 "\n",
+        priv->port, (uint32_t)frame->can_id, frame->can_dlc);
+
+  if (!sam_lin_txready(priv))
+    {
+      canerr("ERROR: transmitter is busy\n");
```

Comment:
Use `nerr` to be consistent with `ninfo`.

Reply:
Done